### PR TITLE
Symbolic shapes mega merge PR (Oct 3)

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86160

- TensorGeometry supports symint
- check_size supports symint
- functorch batch rule improved symint
- Some operator support for symint in LTC
- More supported operations on SymInt and SymFloat
- More symint support in backwards formulas

This merge includes code contributions from bdhirsh and anjali411.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -446,8 +499,8 @@ def _maybe_guard_eq(self, expr: "sympy.Eq") -> None:
             if expr.has(sympy.Mod):
                 mod_expr = tuple(expr.atoms(sympy.Mod))[0]
                 try:
-                    solutions = sympy.solve(lhs - rhs, mod_expr, dict=True)
-                    if len(solutions) == 1 and solutions[0][mod_expr] == 0:
+                    solutions = sympy.solve(lhs - rhs, mod_expr)
```
Comment (Chillee) [inline torch/fx/experimental/symbolic_shapes.py:502]: Don't land this, I think.

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -473,5 +526,6 @@ def evaluate_expr(self, expr: "sympy.Expr"):
         # NB: drop two frames; evaluate_expr and the Sym* function that
         # actually called us
         stack = ''.join(traceback.format_list(traceback.extract_stack()[:-2]))
+        stack += ''.join(torch.utils.get_cpp_backtrace())
```
Comment (Chillee) [inline torch/fx/experimental/symbolic_shapes.py:529]: I think this also slows down tracing quite a bit.

File: torch/fx/experimental/symbolic_shapes.py (hunk 1 above)
Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:502]: I'm going to revert this from branch too then

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -177,6 +178,26 @@ def eval(cls, base, divisor):
                     sympy.simplify(base / gcd), sympy.simplify(divisor / gcd)
                 )
 
+    class Ceil(sympy.Function):
```
Comment (Chillee) [inline torch/fx/experimental/symbolic_shapes.py:181]: Why can't we use this here? https://docs.sympy.org/latest/modules/functions/elementary.html

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:181]: ^- @bdhirsh

**Review by Chillee — APPROVED** [review]
symbolic shapes change is unclear to me, but will approve to unblock.

File: torch/fx/experimental/symbolic_shapes.py (hunk 3 above)
Comment (bdhirsh) [reply torch/fx/experimental/symbolic_shapes.py:181]: Oh yeah we should use ‘ceiling’, I must have missed it. I mostly just copied off of how your did floor div

**Pull request closed without merging**